Get VSCode Java language server working with JDK 17

In devcontainer, currently getting this:
```
WARNING: Using incubator modules: jdk.incubator.foreign, jdk.incubator.vector
[Error - 1:25:52 AM] Jul 10, 2022, 1:25:52 AM An internal error occurred during: "Loading available Gradle versions".
null
java.lang.ExceptionInInitializerError
	at java.base/javax.crypto.Cipher.getInstance(Cipher.java:546)
	at java.base/sun.security.ssl.SSLCipher.isTransformationAvailable(SSLCipher.java:511)
	at java.base/sun.security.ssl.SSLCipher.<init>(SSLCipher.java:500)
	at java.base/sun.security.ssl.SSLCipher.<clinit>(SSLCipher.java:82)
	at java.base/sun.security.ssl.CipherSuite.<clinit>(CipherSuite.java:65)
	at java.base/sun.security.ssl.SSLContextImpl.getApplicableSupportedCipherSuites(SSLContextImpl.java:343)
	at java.base/sun.security.ssl.SSLContextImpl$AbstractTLSContext.<clinit>(SSLContextImpl.java:556)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:375)
	at java.base/java.security.Provider$Service.getImplClass(Provider.java:1933)
	at java.base/java.security.Provider$Service.getDefaultConstructor(Provider.java:1964)
	at java.base/java.security.Provider$Service.newInstanceOf(Provider.java:1878)
	at java.base/java.security.Provider$Service.newInstanceUtil(Provider.java:1886)
	at java.base/java.security.Provider$Service.newInstance(Provider.java:1861)
	at java.base/sun.security.jca.GetInstance.getInstance(GetInstance.java:236)
	at java.base/sun.security.jca.GetInstance.getInstance(GetInstance.java:164)
	at java.base/javax.net.ssl.SSLContext.getInstance(SSLContext.java:185)
	at java.base/javax.net.ssl.SSLContext.getDefault(SSLContext.java:110)
	at java.base/javax.net.ssl.SSLSocketFactory.getDefault(SSLSocketFactory.java:83)
	at java.base/javax.net.ssl.HttpsURLConnection.getDefaultSSLSocketFactory(HttpsURLConnection.java:336)
	at java.base/javax.net.ssl.HttpsURLConnection.<init>(HttpsURLConnection.java:292)
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.<init>(HttpsURLConnectionImpl.java:81)
	at java.base/sun.net.www.protocol.https.Handler.openConnection(Handler.java:62)
	at java.base/sun.net.www.protocol.https.Handler.openConnection(Handler.java:57)
	at java.base/java.net.URL.openConnection(URL.java:1094)
	at org.eclipse.buildship.core.internal.util.gradle.PublishedGradleVersions.downloadVersionInformation(PublishedGradleVersions.java:159)
```
But on macOS host, also having problems, apparently with the language server sending an unrecognized option to the JVM when trying to create the JVM. Could it be Corretto-specific? Or is it version-specific?
```
Caused by: org.gradle.api.GradleException: Unable to start the daemon process.
This problem might be caused by incorrect configuration of the daemon.
For example, an unrecognized jvm option is used.
Please refer to the User Manual chapter on the daemon at https://docs.gradle.org/7.4.2/userguide/gradle_daemon.html
Process command line: /Users/inonit/.inonit/tool/vscode/jdk/Contents/Home/bin/java -XX:MaxPermSize=512m -XX:+HeapDumpOnOutOfMemoryError --add-opens java.base/java.util=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.lang.invoke=ALL-UNNAMED --add-opens java.base/java.util=ALL-UNNAMED --add-opens java.prefs/java.util.prefs=ALL-UNNAMED --add-opens java.prefs/java.util.prefs=ALL-UNNAMED --add-opens java.base/java.nio.charset=ALL-UNNAMED --add-opens java.base/java.net=ALL-UNNAMED --add-opens java.base/java.util.concurrent.atomic=ALL-UNNAMED -Xmx2048M -Dfile.encoding=utf8 -Duser.country=US -Duser.language=en -Duser.variant -cp /Users/inonit/.gradle/wrapper/dists/gradle-7.4.2-bin/48ivgl02cpt2ed3fh9dbalvx8/gradle-7.4.2/lib/gradle-launcher-7.4.2.jar org.gradle.launcher.daemon.bootstrap.GradleDaemon 7.4.2
Please read the following process output to find out more:
-----------------------
Unrecognized VM option 'MaxPermSize=512m'
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```

--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -19,5 +19,8 @@
 		"master"
 	],
 	"java.jdt.ls.vmargs": "-XX:+UseParallelGC -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true -Xmx2G -Xms100m",
+	//	See	https://github.com/redhat-developer/vscode-java/issues/2551 for why the below is currently needed; should be considered
+	//	temporary as presumably at some point that bug will be fixed.
+	"java.import.gradle.jvmArguments": "-XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError -Xms256m -Xmx1g",
 	"task.allowAutomaticTasks": true
 }

--- a/contributor/jrunscript.fifty.ts
+++ b/contributor/jrunscript.fifty.ts
@@ -45,6 +45,7 @@
 			if (hasGit) fifty.load("../rhino/tools/git/module.fifty.ts");
 			fifty.load("../rhino/tools/gcloud/module.fifty.ts");
 			fifty.load("../jsh/launcher/test/suite.fifty.ts");
+			fifty.load("../jsh/test/remote.fifty.ts");
 			fifty.load("../jsh/loader/jsh.fifty.ts");
 			fifty.load("../jsh/script/plugin.jsh.fifty.ts");
 			if (isMkcertImplemented) fifty.load("../jsh/unit/plugin.jsh.web.fifty.ts");

--- a/jsh.bash
+++ b/jsh.bash
@@ -22,6 +22,7 @@ if [ "$0" == "bash" ]; then
 	rmdir ${JSH_LOCAL_JDKS}
 	JDK_USER_JDKS=/dev/null
 	JSH_LAUNCHER_GITHUB_PROTOCOL="${JSH_LAUNCHER_GITHUB_PROTOCOL:-https}"
+	JSH_LAUNCHER_GITHUB_BRANCH="${JSH_LAUNCHER_GITHUB_BRANCH:-master}"
 else
 	JSH_LOCAL_JDKS="${JSH_LOCAL_JDKS:-$(dirname $0)/local/jdk}"
 	JDK_USER_JDKS="${JSH_USER_JDKS:-${HOME}/.slime/jdk}"
@@ -362,7 +363,7 @@ if [ "$0" == "bash" ]; then
 		if [ -n "${JSH_HTTPS_PROXY_HOST}" ]; then
 			CURL_PROXY_ARGUMENTS="${CURL_PROXY_ARGUMENTS} -x ${JSH_LAUNCHER_GITHUB_PROTOCOL}://${JSH_HTTPS_PROXY_HOST}:${JSH_HTTPS_PROXY_PORT}"
 		fi
-		AUTHORIZATION_SCRIPT_URL="${JSH_LAUNCHER_GITHUB_PROTOCOL}://raw.githubusercontent.com/davidpcaldwell/slime/master/rhino/tools/github/${SCRIPT}"
+		AUTHORIZATION_SCRIPT_URL="${JSH_LAUNCHER_GITHUB_PROTOCOL}://raw.githubusercontent.com/davidpcaldwell/slime/${JSH_LAUNCHER_GITHUB_BRANCH}/rhino/tools/github/${SCRIPT}"
 		echo $(curl -L ${CURL_PROXY_ARGUMENTS} -u ${JSH_GITHUB_USER}:${JSH_GITHUB_PASSWORD} ${AUTHORIZATION_SCRIPT_URL})
 	}
 
@@ -370,7 +371,7 @@ if [ "$0" == "bash" ]; then
 	#	AUTHORIZATION_SCRIPT=$(get_authorization_script authorize.js)
 	#	echo ${AUTHORIZATION_SCRIPT}
 	AUTHORIZATION_SCRIPT="//  no-op"
-	${JRUNSCRIPT} ${JSH_NETWORK_ARGUMENTS} -e "${AUTHORIZATION_SCRIPT}" -e "load('${JSH_LAUNCHER_GITHUB_PROTOCOL}://raw.githubusercontent.com/davidpcaldwell/slime/master/rhino/jrunscript/api.js?jsh')" "$@"
+	${JRUNSCRIPT} ${JSH_NETWORK_ARGUMENTS} -e "${AUTHORIZATION_SCRIPT}" -e "load('${JSH_LAUNCHER_GITHUB_PROTOCOL}://raw.githubusercontent.com/davidpcaldwell/slime/${JSH_LAUNCHER_GITHUB_BRANCH}/rhino/jrunscript/api.js?jsh')" "$@"
 else
 	${JRUNSCRIPT} $(dirname $0)/rhino/jrunscript/api.js jsh "$@"
 fi

--- a/jsh/loader/java/inonit/script/jsh/Main.java
+++ b/jsh/loader/java/inonit/script/jsh/Main.java
@@ -56,7 +56,12 @@ public class Main {
 			String host = "raw.githubusercontent.com";
 			if (string.startsWith("http://" + host) || string.startsWith("https://" + host)) {
 				try {
-					if (string.indexOf("//raw.githubusercontent.com/davidpcaldwell/slime/master/") != -1) {
+					String githubUrlPortion = "//raw.githubusercontent.com/davidpcaldwell/slime/";
+					if (string.indexOf(githubUrlPortion) != -1) {
+						int branchIndex = string.indexOf(githubUrlPortion) + githubUrlPortion.length();
+						int branchEnd = string.substring(branchIndex).indexOf("/");
+						String branch = string.substring(branchIndex, branchIndex + branchEnd);
+						System.err.println("branch = " + branch);
 						String protocol = "https";
 						if (System.getenv("JSH_LAUNCHER_GITHUB_PROTOCOL") != null) {
 							protocol = System.getenv("JSH_LAUNCHER_GITHUB_PROTOCOL");
@@ -70,7 +75,7 @@ public class Main {
 								throw new RuntimeException(e);
 							}
 						}
-						return Code.Loader.github(new java.net.URL(protocol + "://github.com/davidpcaldwell/slime/archive/refs/heads/master.zip"), "slime-master");
+						return Code.Loader.github(new java.net.URL(protocol + "://github.com/davidpcaldwell/slime/archive/refs/heads/" + branch + ".zip"), "slime-" + branch);
 						//	Below is older API that "works" in the sense that it requests the correct URLs etc. but is horrendously
 						//	inefficient and switfly hits a GitHub API rate limit
 						//	return Code.Loader.githubApi(new URL(string));

--- a/jsh/loader/jsh.fifty.ts
+++ b/jsh/loader/jsh.fifty.ts
@@ -130,6 +130,7 @@ namespace slime.jsh {
 					}
 				) => any
 			}
+			synchronous: slime.runtime.loader.Exports["synchronous"]
 		}
 
 		(

--- a/jsh/loader/jsh.js
+++ b/jsh/loader/jsh.js
@@ -421,7 +421,8 @@
 							}
 						},
 						//	TODO	check semantics; maybe this returns null on non-existence, currently not documented in jsh/loader/plugin.api.html
-						kotlin: ($slime.getLibraryFile("kotlin") && $slime.getLibraryFile("kotlin").exists()) ? kotlin() : void(0)
+						kotlin: ($slime.getLibraryFile("kotlin") && $slime.getLibraryFile("kotlin").exists()) ? kotlin() : void(0),
+						synchronous: $slime.loader.synchronous
 					}
 				}
 			)();

--- a/jsh/test/remote.fifty.ts
+++ b/jsh/test/remote.fifty.ts
@@ -1,0 +1,86 @@
+//	LICENSE
+//	This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not
+//	distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+//	END LICENSE
+
+namespace slime.jsh.test.remote {
+	(
+		function(
+			fifty: slime.fifty.test.Kit
+		) {
+			const { verify } = fifty;
+			const { $api, jsh } = fifty.global;
+
+			var slime = fifty.jsh.file.relative("../..");
+
+			var loader = jsh.file.world.Location.directory.loader.synchronous({ root: slime });
+
+			var code: {
+				testing: slime.jsh.unit.mock.github.test.Script
+			} = {
+				testing: jsh.loader.synchronous.script("rhino/tools/github/test/module.js")(loader)
+			};
+
+			var library = {
+				testing: code.testing({
+					slime: jsh.file.object.directory(slime)
+				})
+			};
+
+			var lock = jsh.java.Thread.Lock();
+
+			var server = library.testing.startMock(jsh);
+
+			var toInvocation = function(line: string[], input?: string) {
+				return jsh.shell.Invocation.create({
+					command: line[0],
+					arguments: line.slice(1),
+					stdio: {
+						input: input,
+						output: "string",
+						error: "string"
+					}
+				});
+			};
+
+			var getOutput = $api.fp.pipe(
+				$api.fp.impure.tap(function(t) {
+					debugger;
+				}),
+				$api.fp.world.mapping(jsh.shell.world.question),
+				$api.fp.impure.tap(function(t) {
+					debugger;
+				}),
+				$api.fp.property("stdio"),
+				$api.fp.property("output")
+			)
+
+			fifty.tests.suite = function() {
+				var settings: slime.jsh.unit.mock.github.test.Settings = {
+					mock: server,
+					branch: "local"
+				};
+				var download = library.testing.getDownloadJshBashCommand(
+					jsh.shell.PATH,
+					settings
+				);
+				var invoke = library.testing.getBashInvocationCommand(settings);
+				jsh.shell.console(download.join(" "));
+				jsh.shell.console(invoke.join(" "));
+				var launcherBashScript = $api.fp.now.invoke(
+					toInvocation(download),
+					getOutput
+				);
+				jsh.shell.console("bash = " + launcherBashScript);
+				var scriptOutput = $api.fp.now.invoke(
+					toInvocation(invoke, launcherBashScript),
+					getOutput
+				);
+				var output = JSON.parse(scriptOutput);
+				verify(output).evaluate.property("engines").is.type("object");
+			}
+		}
+	//@ts-ignore
+	)(fifty);
+}

--- a/loader/Loader.fifty.ts
+++ b/loader/Loader.fifty.ts
@@ -5,6 +5,12 @@
 //	END LICENSE
 
 namespace slime {
+	export namespace loader {
+		export namespace synchronous {
+			export type Script<C,E> = (c: C) => E
+		}
+	}
+
 	export namespace runtime.loader {
 		export interface Node {
 			name: string
@@ -47,7 +53,7 @@ namespace slime {
 					parent: (path: string[]) => boolean
 				}) => (loader: Synchronous<T>) => Location[]
 
-				script: (path: string) => <T>(loader: Synchronous<T>) => <C,E>(c: C) => E
+				script: (path: string) => <T,C,E>(loader: Synchronous<T>) => slime.loader.synchronous.Script<C,E>
 			}
 		}
 

--- a/rhino/jrunscript/api.js
+++ b/rhino/jrunscript/api.js
@@ -566,7 +566,7 @@
 			/**
 			 *
 			 * @param { slime.jrunscript.native.java.net.URL } url
-			 * @returns { { zip: slime.jrunscript.native.java.net.URL, path: string } }
+			 * @returns { { zip: slime.jrunscript.native.java.net.URL, branch: string, path: string } }
 			 */
 			var toGithubArchiveLocation = function(url) {
 				var string = String(url);
@@ -579,6 +579,7 @@
 					);
 					return {
 						zip: zipurl,
+						branch: githubMatch[2],
 						path: githubMatch[3]
 					}
 				} else {
@@ -668,9 +669,9 @@
 						if (location) {
 							var archive = archives[String(location.zip.toExternalForm())];
 							if (archive) {
-								var _inputStream = archive.read("slime-master/" + location.path);
+								var _inputStream = archive.read("slime-" + location.branch + "/" + location.path);
 								if (!_inputStream) {
-									$api.debug("Not found in " + archive + ": " + "slime-master/" + location.path);
+									$api.debug("Not found in " + archive + ": " + "slime-" + location.branch + "/" + location.path);
 									return null;
 								}
 								return io.readJavaString(_inputStream);
@@ -688,7 +689,7 @@
 							if (archive) {
 								$api.debug("Listing URL " + _url + " path " + location.path);
 								//	TODO	should not hard-code master
-								var list = archive.list("slime-master/" + location.path);
+								var list = archive.list("slime-" + location.branch + "/" + location.path);
 
 								var rv = [];
 								for (var i=0; i<list.length; i++) {

--- a/rhino/tools/github/plugin.jsh.js
+++ b/rhino/tools/github/plugin.jsh.js
@@ -37,6 +37,11 @@
 					}
 					return function(request) {
 						var host = request.headers.value("host");
+
+						//	TODO	relaxing branch/ref requirements so that we always pretend that the requested branch was also
+						//			the current branch and can just serve from the filesystem. We could probably write a more
+						//			complex algorithm but we'd also have to figure out whether it would work in our GitHub Action
+						//			(do we even install git on the server in that action?).
 						if (host == "raw.githubusercontent.com") {
 							var response = authorize(request);
 							if (response) return response;
@@ -49,7 +54,9 @@
 							var branch = repository.branch().filter(function(b) {
 								return b.current;
 							})[0];
-							if (branch.name == ref) {
+							//	TODO	for the purposes of this script, we pretend that master points to local also, because
+							//			when running the shell via jsh.bash
+							if (ref == "local") {
 								var file = repository.directory.getFile(match[4])
 								return (file) ? {
 									status: { code: 200 },
@@ -62,7 +69,7 @@
 									status: { code: 404 }
 								}
 							} else {
-								throw new Error("Unsupported: branch and ref different.");
+								throw new Error("Unsupported: ref not 'local'.");
 							}
 						} else if (host == "api.github.com") {
 							var pattern = /^repos\/(.*)\/(.*)\/contents\/(.*)$/;
@@ -72,7 +79,7 @@
 								var repo = match[2];
 								var path = match[3];
 								//	TODO	can take ref as query parameter; see https://developer.github.com/v3/repos/contents/
-								var ref = "master";
+								var ref = "local";
 								var repository = o.src[user][repo];
 								var branch = repository.branch().filter(function(b) {
 									return b.current;
@@ -126,7 +133,7 @@
 								var user = match[1];
 								var repo = match[2];
 								var version = match[3];
-								if (user == "davidpcaldwell" && repo == "slime" && version == "master") {
+								if (user == "davidpcaldwell" && repo == "slime" && version == "local") {
 									//	create zip file of source tree, excluding .git
 									var isTopLevel = function(node) {
 										return node.parent.toString() == jsh.shell.jsh.src.toString();

--- a/rhino/tools/github/test/module.fifty.ts
+++ b/rhino/tools/github/test/module.fifty.ts
@@ -11,6 +11,7 @@ namespace slime.jsh.unit.mock.github.test {
 
 	export interface Settings {
 		mock?: slime.jsh.unit.mock.Web
+		branch?: string
 		optimize?: boolean
 		debug?: boolean
 		token?: () => string
@@ -18,6 +19,10 @@ namespace slime.jsh.unit.mock.github.test {
 
 	export interface Exports {
 		startMock: (jsh: slime.jsh.Global) => slime.jsh.unit.mock.Web
+
+		getDownloadJshBashCommand: (PATH: slime.jrunscript.file.Searchpath, options: Pick<slime.jsh.unit.mock.github.test.Settings,"mock" | "token" | "branch">) => string[]
+
+		getBashInvocationCommand: (options: slime.jsh.unit.mock.github.test.Settings) => string[]
 
 		/**
 		 * Outputs a single string, suitable for use at the shell command line, that will invoke a shell with the given settings
@@ -37,5 +42,5 @@ namespace slime.jsh.unit.mock.github.test {
 	//@ts-ignore
 	)(fifty);
 
-	export type Script = slime.loader.Script<Context,Exports>
+	export type Script = slime.loader.synchronous.Script<Context,Exports>
 }

--- a/rhino/tools/github/test/module.js
+++ b/rhino/tools/github/test/module.js
@@ -42,13 +42,15 @@
 
 		/**
 		 * @param { slime.jrunscript.file.Searchpath } PATH
-		 * @param { Pick<slime.jsh.unit.mock.github.test.Settings,"mock" | "token"> } p
+		 * @param { Pick<slime.jsh.unit.mock.github.test.Settings,"mock" | "token" | "branch"> } p
 		 * @returns
 		 */
 		var getDownloadJshBashCommand = function(PATH,p) {
 			/** @type { string[] } */
 			var command = [];
 			var PROTOCOL = (p.mock) ? "http" : "https";
+			var branch = p.branch || "master";
+			var URL = PROTOCOL + "://raw.githubusercontent.com/davidpcaldwell/slime/" + branch + "/jsh.bash";
 			if (PATH.getCommand("curl")) {
 				command.push("curl", "-v");
 				if (p.token) {
@@ -58,7 +60,7 @@
 					command.push("--proxy", "https://127.0.0.1:" + p.mock.https.port);
 				}
 				command.push("-L");
-				command.push(PROTOCOL + "://raw.githubusercontent.com/davidpcaldwell/slime/master/jsh.bash");
+				command.push(URL);
 				return command;
 			} else if (PATH.getCommand("wget")) {
 				command.push("wget");
@@ -66,7 +68,7 @@
 					command.push("--http-user=" + "davidpcaldwell");
 					command.push("--http-password=" + p.token());
 				}
-				command.push(PROTOCOL + "://raw.githubusercontent.com/davidpcaldwell/slime/master/jsh.bash");
+				command.push(URL);
 				if (p.mock) {
 					command.push("-e", "use_proxy=yes");
 					command.push("-e", "http_proxy=" + "http://127.0.0.1:" + p.mock.port);
@@ -87,6 +89,7 @@
 			/** @type { string[] } */
 			var command = [];
 			var PROTOCOL = "https";
+			var branch = p.branch || "master";
 			if (p.mock) {
 				command.push("env");
 				command.push("JSH_HTTP_PROXY_HOST=127.0.0.1", "JSH_HTTP_PROXY_PORT=" + p.mock.port);
@@ -104,8 +107,14 @@
 			if (p.token) {
 				command.push("JSH_GITHUB_USER=davidpcaldwell", "JSH_GITHUB_PASSWORD=" + p.token());
 			}
+			if (p && p.branch) command.push("JSH_LAUNCHER_GITHUB_BRANCH=" + p.branch);
+			//	TODO	for now, we simply try to force-disable the Rhino debugger (because by default we will be running in
+			//			Nashorn), even if this script is running in the Rhino debugger. Probably could do more here to think this
+			//			through, including allowing the running of remote shells under Rhino, but bootstrapping with Nashorn may
+			//			be fine.
+			command.push("JSH_DEBUG_SCRIPT=");
 			command.push("bash", "-s");
-			command.push(PROTOCOL + "://raw.githubusercontent.com/davidpcaldwell/slime/master/jsh/test/jsh-data.jsh.js");
+			command.push(PROTOCOL + "://raw.githubusercontent.com/davidpcaldwell/slime/" + branch + "/jsh/test/jsh-data.jsh.js");
 			return command;
 		}
 
@@ -124,6 +133,8 @@
 
 		$export({
 			startMock: startMock,
+			getDownloadJshBashCommand: getDownloadJshBashCommand,
+			getBashInvocationCommand: getBashInvocationCommand,
 			getCommandLine: getCommand
 		})
 	}

